torch/distributed/tensor/parallel/loss_parallel.py

```diff
@@ -0,0 +1,402 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates
+import contextlib
+from typing import cast, Dict, Optional, Tuple
+
+import torch
+import torch._prims_common as utils
+import torch.distributed._functional_collectives as funcol
+import torch.distributed.distributed_c10d as c10d
+from torch import Tensor
+from torch.distributed._tensor import DTensor, Replicate
+from torch.distributed._tensor.ops.embedding_ops import _MaskPartial
+from torch.distributed._tensor.ops.math_ops import (
+    _skip_dim,
+    Reduction,
+    replicate_reduction_dims,
+)
+from torch.distributed._tensor.placement_types import Placement, TensorMeta
+from torch.distributed.device_mesh import DeviceMesh
+
+aten = torch.ops.aten
+
+
+@contextlib.contextmanager
+def loss_parallel():
```

Comment:
Re 1: The semantic of `cross_entropy` says the `vocab_dim` should always be 1 if input dim >= 2 and 0 otherwise. Currently I'm throwing exception if input is not sharded on `vocab_dim`, in which case there is no need to enable this context manager.

Re 2: Return torch.Tensor sounds easy, can we just don't wrap the results to DTensor in the last backward handler? I'm not sure how to turn inputs from torch.Tensor without user doing anything extra, since at least one input has to be DTensor to route into DTensor dispatcher?